TextInput: Fix stale cursor/anchor offsets and undo panic when text changes externally

tronical opened the pull request:
When `text` is assigned programmatically, the cursor and anchor byte offsets were left stale (reporting wrong values, #331) and the undo/redo stack kept positions into the replaced text, causing an out-of-bounds panic on undo (#9024).

`TextInput` now keeps an internal mirror of the text and installs a change handler on the `text` property. On an external assignment it re-clamps the cursor and anchor offsets to the new text and clears the undo/redo history; internal edits route through a helper that keeps the mirror in sync so they aren't mistaken for external changes. The realignment also emits `cursor-position-changed`.

Fixes #331
Fixes #9024

ogoffart commented:
> This is still missing the bit about emitting cursor-position-changed

Maybe we need a `changed xxx` in the implementation of the LineEditBase/TextEditBase ?

tronical commented:
I was thinking the same, but OTOH the callback is public API in TextInput and it was nice if it worked also when changing text.

ogoffart commented:
Then we could perhaps have a PropertyTracker or ChangedTracker in the TextInput  (and we wouldn't need the proxy bindings)

ogoffart commented on internal/core/items/text.rs:
```diff
@@ -487,8 +487,13 @@ pub struct TextInput {
     pub cursor_position_byte_offset: Property<i32>,
     pub anchor_position_byte_offset: Property<i32>,
+    pub unchecked_cursor_position_byte_offset: Property<i32>,
+    pub unchecked_anchor_position_byte_offset: Property<i32>,
```
We could have a `#[repr(C)] struct CursorPosition { cursor_pos: i32, anchor: i32 }` ` 
then we can put that in a Property.
(But doesn't have to be in this PR, this is just a thought)

tronical commented:
Discussed this a bit further with Olivier, also in light of #9024, which suffers from the same problem.

We concluded that `TextInput` needs a copy of the text, to which we align all other data structures such as the anchor/cursor offsets as well as the undo-redo stack. For the public text property, we install a change handler and whenever that's triggered, we update the internal text and all associated data structures.

tronical committed:
TextInput: Realign cursor/anchor and undo stack when text changes externally

Setting the `text` property programmatically left the cursor and anchor
byte offsets stale and the undo/redo stack pointing into the replaced text,
so the offsets reported wrong values and undoing could panic.

Keep an internal mirror of the text and install a change handler that, on an
external assignment, re-clamps the cursor/anchor offsets to the new text and
clears the undo/redo history. Internal edits route through a helper that keeps
the mirror in sync so they are not mistaken for external changes.

ChangeLog: TextInput: Setting `text` programmatically now re-clamps the cursor/anchor offsets and resets the undo history, fixing stale cursor-position values and a panic when undoing after the text was replaced.

Fixes #331
Fixes #9024

ogoffart commented on internal/core/items/text.rs:
```diff
@@ -771,10 +771,62 @@ pub struct TextInput {
+}
+
+#[derive(Default)]
+pub struct TextInputData {
```
Why do these needs to be in a Box? Both SharedString and ChangeTracker can be inline, i think.  (although i might be wrong)

tronical replied:
`ChangeTracker` implements `Default` and `Drop` for Rust, both absent from C++. Want me to wrap that in C++?

tronical replied:
I stand corrected, it's already there :)

ogoffart commented on tests/cases/text/issue_9024_undo_after_text_set.slint:
````diff
@@ -0,0 +1,70 @@
+instance.invoke_do_undo();
+assert_eq!(instance.get_test_text(), "");
+```
+*/
````
can we also have a C++ coverage?

ogoffart reviewed (approved):
Nice.

I think we would benefit from even more tests.  (With, for example, emoji or unicode things with cursor in the middle and things like that)

tronical committed:
TextInput: Inline the internal data instead of boxing it

tronical committed:
TextInput: Add C++ coverage for the issue #9024 regression test

tronical committed:
TextInput: Snap external-text realignment to grapheme boundaries
